### PR TITLE
Test Enumerable Methods Using Rspec

### PR DESCRIPTION
**Pull request by AndresFMoya**
Hi @bolah2009. **Thanks for your feedback**. For this Pull Request, these are the changes made:

- `Readme` Update describing the objectives of the project and the results of the tests performed.
- Integration of RSPec 3.8 to Test `Enumerable` Module.
- The `enumerable_rspec` file was created and tests were performed on the the `Enumerable` module covering the following cases in the methods listed:

```
Enumerable
  When the block returns true to every element
    #my_all
      should returns true if the block never returns false or nil
  When no block is given
    #my_any
      should return true if at least one of the collection members is not false or nil
  When no block is given
    #my_each
      should returns an Enumerator
  When no block is given
    #my_none
      return true only if none of the collection members is true
  When block is given
    #my_each_with_index
      should Calls block with two arguments, the item and its index, for each item in enum
  When the block returns false to any element
    #my_all?
      should returns true if the block never returns false or nil
  When block is given
    #my_select
      should returns an array containing all elements of enum for which the given block returns a true value.
  When the block returns false to any element
    #my_none?
      should returns true if the block never returns true for all elements
  When block returns true to any element
    #my_any?
      should returns true if the block ever returns a value other than false or nil.
  When no block is given
    #my_each_with_index
      should returns an Enumerator
  When the block returns true to any elemend
    #my_none?
      should returns false if the block never returns true for all elements
  When block is no given
    #my_select
      should returns an Enumerator
  When block is given
    #my_each
      should iterate from every item on the array
  When block returns false to every element
    #my_any?
      should returns false if the block ever returns false or nil.

Finished in 0.01017 seconds (files took 0.14339 seconds to load)
14 examples, 0 failures
```

Regards!

### CODE REVIEW

**Review by Oghenebrume50 — CHANGES_REQUESTED** [review]
Hi, Team, this is a really good written test :clap: the only thing left to do now is to use make sure negative and positive scenarios are covered, this is done for cases like my_each and my_each_with_index but the other's aren't covered.

Basically positive and negative means that, if you test for the truthy outcome of a method you want to also make a test for the falsy outcome.

**Comment by AndresFMoya** [comment]
Hi @Oghenebrume50. Thanks for your feedback. We add negative test cases to the remaining methods. Now we have 12 test cases. 2 for every method. We sent a new code review request. Regards!

File: spec/enumerable_spec.rb [hunk 1]
```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+require_relative '../enumerable.rb'
+
+RSpec.describe Enumerable do
+  arr = (1..10).to_a
```
Comment (bolah2009) [inline spec/enumerable_spec.rb]: Consider using `let` helper here instead.
> Use let to define a memoized helper method. The value will be cached
across multiple calls in the same example but not across examples.

```suggestion
  let(:arr){(1..10).to_a}
```

#### Learn more about [`let` here](https://relishapp.com/rspec/rspec-core/v/2-5/docs/helper-methods/let-and-let).

File: spec/enumerable_spec.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  context 'When block is no given' do
+    describe '#my_select' do
+      it 'should returns an Enumerator' do
+        ans = arr.my_select
+        expect(ans).to be_an Enumerator
+      end
+    end
+  end
+
+  context 'When the block returns true to every element' do
```
Comment (bolah2009) [inline spec/enumerable_spec.rb:64]: What happens when no block is given?

```ruby
  def my_any?(arg = nil)
    if block_given?
      my_each { |item| return true if yield(item) }
    # elsif arg.nil?
    #   my_each { |item| return true if item }
    # else
    #   my_each { |item| return true if item.class == arg }
    end
    false
  end
```

Currently, the above method passes the test. Please, let's test for the commented out cases.
- when an argument is not passed and no block is given.
- when an argument is passed and no block is given.

File: spec/enumerable_spec.rb [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    describe '#my_any?' do
+      it 'should returns false if the block ever returns false or nil.' do
+        ans = arr.my_any? { |number| number > 12 }
+        expect(ans).to eql(false)
+      end
+    end
+  end
+
+  context 'When the block returns false to any element' do
+    describe '#my_none?' do
```
Comment (bolah2009) [inline spec/enumerable_spec.rb:110]: What happens when no block is given? :thinking: 

```ruby
  def my_none?(&block)
    result = true
    if block
      my_each { |e| result = false if yield(e) }
    # else
    #   my_each { |e| result = false if e }
    end
    result
  end
```

Currently, the above method passes the test. Please, let's test for the commented out cases.
- when no block is given.

**Review by bolah2009 — CHANGES_REQUESTED** [review]
# Awesome job @AndresFMoya 

### Almost there :smile: 

Let's cover more cases in the test for the following methods:
- `my_any?`
- `my_none?`

- Let's make this PR (Pull Request) look more professional. [See below]()
  - [REQUIRED] Use meaningful names for  Pull Requests.
  - [REQUIRED] Add a short summary of changes in Pull Request description.

#### :recycle:  Make these changes and submit another code review request. :+1: 

#### Feel free to reach me on slack if you having issues implementing these features. :+1: 

Good luck with your next submission. :star2: 

Cheers
[Bola](https://bolabuari.com/)

###### PS: Note that comments tagged as `OPTIONAL` are not part of the minimum requirements to approve this PR

File: spec/enumerable_spec.rb (hunk 1 above)
Comment (rOluochKe) [reply spec/enumerable_spec.rb]: resolved

File: spec/enumerable_spec.rb (hunk 3 above)
Comment (rOluochKe) [reply spec/enumerable_spec.rb:110]: added when no block is given

File: spec/enumerable_spec.rb (hunk 2 above)
Comment (rOluochKe) [reply spec/enumerable_spec.rb:64]: done - when an argument is not passed and no block is given
done - when an argument is passed and no block is given.

**Comment by Torres-ssf** [comment]
### Pull Request Under Review By Sérgio (TSE)

**Review by Torres-ssf — APPROVED** [review]
### Hi there team :hand: 

Great job implementing the required changes :clap: 
### I'm approving this project :+1: 

Good luck with the next project. Happy coding :computer:

**Pull request merged**